rust/agama-server/src/users/web.rs

```diff
@@ -0,0 +1,253 @@
+            Box::pin(first_user_changed_stream(dbus.clone()).await?),
+        ),
+        (
+            ROOT_PASSWORD_ID,
+            Box::pin(root_password_changed_stream(dbus.clone()).await?),
+        ),
+        (
+            ROOT_SSHKEY_ID,
+            Box::pin(root_ssh_key_changed_stream(dbus.clone()).await?),
+        ),
+    ];
+
+    Ok(result)
+}
+
+async fn first_user_changed_stream(
+    dbus: zbus::Connection,
+) -> Result<impl Stream<Item = Event> + Send, Error> {
+    let proxy = Users1Proxy::new(&dbus).await?;
+    let stream = proxy
+        .receive_first_user_changed()
+        .await
+        .then(|change| async move {
+            if let Ok(user) = change.get().await {
+                let user_struct = FirstUser {
+                    full_name: user.0,
+                    user_name: user.1,
+                    password: user.2,
+                    autologin: user.3,
+                    data: user.4,
+                };
+                return Some(Event::FirstUserChanged(user_struct));
+            }
+            None
+        })
+        .filter_map(|e| e);
+    Ok(stream)
+}
+
+async fn root_password_changed_stream(
+    dbus: zbus::Connection,
+) -> Result<impl Stream<Item = Event> + Send, Error> {
+    let proxy = Users1Proxy::new(&dbus).await?;
+    let stream = proxy
+        .receive_root_password_set_changed()
+        .await
+        .then(|change| async move {
+            if let Ok(is_set) = change.get().await {
+                return Some(Event::RootPasswordChanged {
+                    password_is_set: is_set,
+                });
+            }
+            None
+        })
+        .filter_map(|e| e);
+    Ok(stream)
+}
+
+async fn root_ssh_key_changed_stream(
+    dbus: zbus::Connection,
+) -> Result<impl Stream<Item = Event> + Send, Error> {
+    let proxy = Users1Proxy::new(&dbus).await?;
+    let stream = proxy
+        .receive_root_sshkey_changed()
+        .await
+        .then(|change| async move {
+            if let Ok(key) = change.get().await {
+                let value = if key.is_empty() { None } else { Some(key) };
+                return Some(Event::RootSSHKeyChanged { key: value });
+            }
+            None
+        })
+        .filter_map(|e| e);
+    Ok(stream)
+}
+
+/// Sets up and returns the axum service for the software module.
+pub async fn users_service(dbus: zbus::Connection) -> Result<Router, ServiceError> {
+    const DBUS_SERVICE: &str = "org.opensuse.Agama.Manager1";
+    const DBUS_PATH: &str = "/org/opensuse/Agama/Users1";
+
+    let users = UsersClient::new(dbus.clone()).await?;
+    let state = UsersState { users };
+    let validation_router = validation_router(&dbus, DBUS_SERVICE, DBUS_PATH).await?;
+    let status_router = service_status_router(&dbus, DBUS_SERVICE, DBUS_PATH).await?;
+    let router = Router::new()
+        .route("/config", get(get_info))
+        .route("/user", put(set_first_user).delete(remove_first_user))
+        .route(
```

Comment:
I think that having a single `/root` resource would be more straightforward (and extensible). Let's say that, tomorrow, we want to add another attribute. 

We could have the following routes (just an idea):

* GET `/root` to get the full user.
* PUT (or PATCH) `/root` to update any of the attributes (just specify the one that changes).
* DELETE `/root/password` to delete the password.
* DELETE `/root/sshkey` to delete the SSH key.

Alternatively, we could use `PUT` to replace the whole resource and `PATCH` to update only one of the elements.

Reply:
yeap, that is something I mention when I speak about need to have some guide to have consistent API. Both ways is possible. When talking about single resource I even think about having `/users/config` and methods get and patch. ( put is problematic as root password get is boolean, but patch needs value and encryption ).